network/benchmark/src/lib.rs

```diff
@@ -0,0 +1,605 @@
+        },
+    }
+}
+
+/// handle work split out by source_loop()
+async fn handler_thread(
+    node_config: NodeConfig,
+    network_client: NetworkClient<BenchmarkMessage>,
+    work_rx: async_channel::Receiver<(NetworkId, Event<BenchmarkMessage>)>,
+    time_service: TimeService,
+    shared: Arc<RwLock<BenchmarkSharedState>>,
+) {
+    let config = node_config.benchmark.unwrap();
+    loop {
+        let (network_id, event) = match work_rx.recv().await {
+            Ok(v) => v,
+            Err(_) => {
+                // RecvError means source was closed, we're done here.
+                return;
+            },
+        };
+        match event {
+            Event::Message(peer_id, wat) => {
+                let msg_wrapper: BenchmarkMessage = wat;
+                handle_direct(
+                    &network_client,
+                    network_id,
+                    peer_id,
+                    msg_wrapper,
+                    time_service.clone(),
+                    shared.clone(),
+                )
+                .await;
+            },
+            Event::RpcRequest(peer_id, msg_wrapper, protocol_id, sender) => {
+                handle_rpc(
+                    peer_id,
+                    msg_wrapper,
+                    protocol_id,
+                    time_service.clone(),
+                    sender,
+                )
+                .await;
+            },
+            Event::NewPeer(wat) => {
+                if config.enable_direct_send_testing {
+                    Handle::current().spawn(direct_sender(
+                        node_config.clone(),
+                        network_client.clone(),
+                        time_service.clone(),
+                        network_id,
+                        wat.remote_peer_id,
+                        shared.clone(),
+                    ));
+                }
+                if config.enable_rpc_testing {
+                    Handle::current().spawn(rpc_sender(
+                        node_config.clone(),
+                        network_client.clone(),
+                        time_service.clone(),
+                        network_id,
+                        wat.remote_peer_id,
+                        shared.clone(),
+                    ));
+                }
+            },
+            Event::LostPeer(_) => {}, // don't care
+        }
+    }
+}
+
+/// run_benchmark_service() does not return, it should be called by .spawn()
+pub async fn run_benchmark_service(
+    node_config: NodeConfig,
+    network_client: NetworkClient<BenchmarkMessage>,
+    network_requests: NetworkServiceEvents<BenchmarkMessage>,
+    time_service: TimeService,
+    shared: Arc<RwLock<BenchmarkSharedState>>,
+) {
+    let config = node_config.benchmark.unwrap();
+    let benchmark_service_threads = config.benchmark_service_threads;
+    let num_threads = match benchmark_service_threads {
+        Some(x) => x,
+        None => match std::thread::available_parallelism() {
```

Comment:
FYI: some folks also tend to use `num_cpus: https://github.com/aptos-labs/aptos-core/blob/216c346f7ebccc6ce82a3f5ad945fff2a23a1c36/aptos-move/aptos-transaction-benchmarks/src/main.rs#L90